### ISSUE TITLE
Suspicious gcc warnings

### ISSUE THREAD

**Issue by agraef** [plugdata-team/plugdata#1037]
Here is the full buildog, grabbed from a build of the latest HEAD this morning: [buildlog.txt](https://github.com/plugdata-team/plugdata/files/12273486/buildlog.txt)

You might wonder, why bother with warnings? Because all the noise can easily drown out the important messages from the compiler. And there _is_ some funky stuff in there (potential bugs) that are worth looking into, along with a slew of JUCE warnings about deprecated methods and such. I give a brief rundown of the interesting ones below.

I already reported this one [inline](https://github.com/plugdata-team/plugdata/commit/4732f349b4be536ca19ca310fa33efa957bc08e4#r123783841):

~~~
/home/ag/Sources/aur4/pd+friends/plugdata-git/src/PlugData/Libraries/libpd/x_libpd_mod_utils.c: In function ‘libpd_creategraphonparent’:
/home/ag/Sources/aur4/pd+friends/plugdata-git/src/PlugData/Libraries/libpd/x_libpd_mod_utils.c:524:39: warning: unsigned conversion from ‘int’ to ‘unsigned char:1’ changes value from ‘2’ to ‘0’ [-Woverflow]
  524 |     ((t_glist*)result)->gl_hidetext = 2;
      |                                       ^
~~~

This is from an older commit, see rev. 4732f349b4be536ca19ca310fa33efa957bc08e4. This is most certainly a bug. Not trying to be annoying, but if you assign 2 to a 1 bit field then you're probably not getting what you want. ;-)

**[Fixed in #1038]** The next one is probably harmless, but might warrant an explicit cast if only to silence the compiler (there are a couple of these):

~~~
[  3%] Building CXX object CMakeFiles/plugdata_core.dir/Source/Objects/ObjectBase.cpp.o
In file included from /home/ag/Sources/github/plugdata/Source/Objects/ObjectBase.cpp:50:
/home/ag/Sources/github/plugdata/Source/Objects/KeyboardObject.h: In member function ‘virtual void KeyboardObject::valueChanged(juce::Value&)’:
/home/ag/Sources/github/plugdata/Source/Objects/KeyboardObject.h:386:62: warning: narrowing conversion of ‘toggle’ from ‘int’ to ‘float’ [-Wnarrowing]
  386 |                 pd->sendDirectMessage(obj.get(), "toggle", { toggle });
      |                                                              ^~~~~~
~~~

This one is from the Ofelia source:

~~~
[ 45%] Building CXX object Libraries/CMakeFiles/externals.dir/plugdata-ofelia/Source/Shared/TcpSocket.cpp.o
/home/ag/Sources/github/plugdata/Libraries/plugdata-ofelia/Source/Shared/TcpSocket.cpp:10:9: warning: #pragma once in main file
   10 | #pragma once
      |         ^~~~
~~~

I'm not sure what the once pragma is supposed to do there, isn't that pragma intended for header files?

Then there's a bunch of warnings in @porres' stuff (cyclone and ELSE). The most suspicious are these (in pd-else): **[Fixed in fbc150f2e2a819b631eda6105e39443714a7d5fa]**

~~~
[ 15%] Building C object Libraries/CMakeFiles/externals.dir/pd-else/Code_source/Compiled/signal/bl.imp2~.c.o
/home/ag/Sources/github/plugdata/Libraries/pd-else/Code_source/Compiled/signal/bl.imp2~.c: In function ‘complex_div_f’:
/home/ag/Sources/github/plugdata/Libraries/pd-else/Code_source/Compiled/signal/bl.imp2~.c:90:29: warning: excess elements in scalar initializer
   90 |     return (t_complex){real,imag};
      |                             ^~~~
~~~

There's a large number of these. That's not how to initialize a `double complex` value, with the above you will only be getting the real part with the imaginary part being ignored and thus set to zero. Assuming ISO C99, with the complex.h header included, the right way to do this is to use something like `CMPLX(real,imag)` (the CMPLX macro is defined in complex.h).

**[Fixed in #1038]** This is from cyclone:

~~~
[ 36%] Building C object Libraries/CMakeFiles/externals.dir/cyclone/cyclone_objects/binaries/control/cycle.c.o
/home/ag/Sources/github/plugdata/Libraries/cyclone/cyclone_objects/binaries/control/cycle.c: In function ‘cycle_anything’:
/home/ag/Sources/github/plugdata/Libraries/cyclone/cyclone_objects/binaries/control/cycle.c:87:16: warning: comparison of distinct pointer types lacks a cast
   87 |     if (s && s != &s)
      |                ^~
~~~

I don't think that this makes much sense. You're comparing a `t_symbol*` with a `t_symbol**` there, so, unless I'm missing something very non-obvious (which at least warrants a cast along with a lengthy comment explaining why this test makes any sense), the `s != &s` test is unlikely to ever yield false unless Pd's internal symbol table is severely messed up. So your test is effectively `s && 1` which is just `s`. Then why not write it that way and make gcc happy?

**[Fixed in #1038]** Then there are a bunch of missing `const` qualifiers:

~~~
[ 38%] Building C object Libraries/CMakeFiles/externals.dir/cyclone/cyclone_objects/binaries/control/spell.c.o
/home/ag/Sources/github/plugdata/Libraries/cyclone/cyclone_objects/binaries/control/spell.c: In function ‘spell_symbol’:
/home/ag/Sources/github/plugdata/Libraries/cyclone/cyclone_objects/binaries/control/spell.c:55:19: warning: passing argument 2 of ‘spell_out’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   55 |     spell_out(x, s->s_name, 1);
      |                  ~^~~~~~~~
~~~

Add `const` to the 2nd `char *ptr` argument of `spell_out()` and gcc will be happy.

**[Fixed in #1038]** And again here:

~~~
[ 39%] Building C object Libraries/CMakeFiles/externals.dir/cyclone/cyclone_objects/binaries/control/tosymbol.c.o
/home/ag/Sources/github/plugdata/Libraries/cyclone/cyclone_objects/binaries/control/tosymbol.c: In function ‘tosymbol_parse’:
/home/ag/Sources/github/plugdata/Libraries/cyclone/cyclone_objects/binaries/control/tosymbol.c:83:27: warning: initialization discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   83 |         char *sepstring = (separator ?
      |                           ^
~~~

**[Fixed in #1038]** And here:

~~~
[ 40%] Building C object Libraries/CMakeFiles/externals.dir/cyclone/shared/control/gui.c.o
/home/ag/Sources/github/plugdata/Libraries/cyclone/shared/control/gui.c: In function ‘hammergui_setup’:
/home/ag/Sources/github/plugdata/Libraries/cyclone/shared/control/gui.c:252:23: warning: initialization discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  252 |         char *cname = class_getname(*ps_hashhammergui->s_thing);
      |                       ^~~~~~~~~~~~~
~~~

**[Fixed in #1038]** And then gcc flags these bad `\:` escape sequences:

~~~
[ 40%] Building C object Libraries/CMakeFiles/externals.dir/cyclone/cyclone_objects/binaries/cyclone_lib.c.o
/home/ag/Sources/github/plugdata/Libraries/cyclone/cyclone_objects/binaries/cyclone_lib.c: In function ‘print_cyclone’:
/home/ag/Sources/github/plugdata/Libraries/cyclone/cyclone_objects/binaries/cyclone_lib.c:648:89: warning: unknown escape sequence: '\:'
  648 |     post(":: Fred Jan Kraan, Alexandre Porres, Derek Kwan, Matt Barber\n\:: and others.");
      |                                                                                         ^
~~~

More like these follow.

There might be more I missed at a cursory glance. The full buildlog is linked at the beginning of this report, so that you can check for yourself.

**Comment by agraef** [comment]
As promised, #1038 fixes all the trivial warnings, so the build log already has much less noise.

Serious stuff (real bugs) left to be fixed at this point are https://github.com/plugdata-team/plugdata/commit/4732f349b4be536ca19ca310fa33efa957bc08e4 and the garbled complex value initializers in pd-else/Code_source/Compiled/signal/bl.imp~.c and bl.imp2~.c.

**Comment by agraef** [comment]
Ok, I think I fixed the complex initializers issue now, see https://github.com/porres/pd-else/pull/1728. The main problem was that the CMPLX macro isn't in the Windows complex.h header, so that I had to fudge it. But all seems to be good now.

Tim, maybe you can pull this over from @porres' upstream where I submitted the PR. While you're at it, can you please sync https://github.com/timothyschoen/pd-else with https://github.com/porres/pd-else again, so that you have the latest CI fixes from there? This should make the Linux and Windows builds over at https://github.com/timothyschoen/pd-else work again.

Talking about CI, can you please make the Windows build on the develop branch work again, too? The breakage started with https://github.com/plugdata-team/plugdata/commit/2bf0dc2e354a29e26a2d545e6cb11d5aabb5c9c2 which, ironically, is titled "Windows fixes". :wink: You need to revert the change to Libraries/CMakeLists.txt. You may wish that alloca() was available on Windows, but even all those wishes will not make it come true. :wink:

**Comment by agraef** [comment]
> You may wish that alloca() was available on Windows, but even all those wishes will not make it come true.

Or maybe it is, but it's called [_alloca](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/alloca?view=msvc-170&redirectedfrom=MSDN). But this will probably require changes to some headers, you can't just call it as alloca() in migw AFAICT.

**Comment by timothyschoen** [comment]
> Or maybe it is, but it's called [_alloca](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/alloca?view=msvc-170&redirectedfrom=MSDN). But this will probably require changes to some headers, you can't just call it as alloca() in mingw AFAICT.

We are not using mingw, we are using MSVC. alloca does exist, but I need to include <malloc.h>.

**Comment by agraef** [comment]
Oops, right.

**Comment by agraef** [comment]
The latest Windows build is looking good as far as the bl.imp* stuff is concerned, too, so it appears that https://github.com/porres/pd-else/pull/1728 works alright MSVC. I'm sure glad I didn't botch that one. :)

There's a bunch of suspicious pointer truncation warnings in the Windows 64 build, though. (A cast to unsigned long doesn't work on 64 bit Windows, you'd have to cast to something like uintptr_t to avoid the truncation issue.)

**Comment by agraef** [comment]
@timothyschoen @porres The truncating pointer casts all seem to come from sprintfs in pd-else, you can see them in the upstream mingw64 build logs, too. Here is a typical example from https://github.com/porres/pd-else/actions/runs/5801848159/job/15727003547: 

~~~
Code_source/Compiled/signal/oscope~.c: In function 'scope_draw_all':
Code_source/Compiled/signal/oscope~.c:334:43: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
  334 |     sprintf(sh->h_pathname, ".x%lx.h%lx", (unsigned long)x->x_cv, (unsigned long)sh);
      |                                           ^
~~~

So they're unlikely to produce immediate crashes. But most of those sprintf strings are probably being used as Tk widget names (as in the above example), so you run the risk of getting name collisions on the Tk side (since the 64 bit pointer values are projected to 32 bit). The risk is probably low (FLW), but if you're getting strange bugs of mixed-up Tk widgets in pd-else on 64 bit Windows only, then this is the likely cause.

This is a long-standing 64 bit Windows idiosyncrasy where `long int` is 32 bit, whereas it's 64 bit on all other 64 bit platforms I know of. Of course the MS engineers had to do their own thing. :) Since mingw64 uses the MS CRT, it has the same wart. It's possible to fix these sprintfs in a portable manner, but it's messy. I just remembered now that I've done a lot of that for purr-data in https://github.com/agraef/purr-data/commit/94c62e36eea53a51b94e978a99ce8d7b33620c3b, but that's for mingw, and I don't know whether there's a similar solution for MSVC.

So the TLDR is that we should probably ignore these for now, and hope for the best. ;-)

**Comment by agraef** [comment]
Tim, it seems that the last remaining issue from this ticket is the wrong `gl_hidetext` value in https://github.com/plugdata-team/plugdata/commit/4732f349b4be536ca19ca310fa33efa957bc08e4. If you have an idea how to fix this (maybe just set the bit to 1 instead of 2?) then we might close this ticket.

**Comment by timothyschoen** [comment]
Done, with 1834b00. I also fixed all the deprecation warnings: 287d37e


> This is a long-standing 64 bit Windows idiosyncrasy where `long int` is 32 bit, whereas it's 64 bit on all other 64 bit platforms I know of.

Heh, @porres and I ran into this problem on Windows recently too: d6806517e504631a7cf21ee51194ff8266fe400b

**Comment by agraef** [comment]
Thanks! Yeah, those "long" ints on 64 bit Windows are a major nuisance, I've run into them in many places as well. ;-)
